workflow-cli panics when issuing command not supported by server API

When using workflow-cli v2.4.0 against a workflow v2.3.0 cluster I encountered a `panic: runtime error: invalid memory address or nil pointer dereference` when attempting to issue commands using any of the features introduced after v2.3.0 including `deis maintenance` and `deis routing`

I would expect a response of "Feature xxx is not supported by server API vM.m.p" not a runtime error.

Here are a couple examples of the errors:

routing top-level
```
$ deis routing
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4bc0cf]

goroutine 1 [running]:
panic(0x8ef380, 0xc82000e100)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/deis/workflow-cli/cmd.DeisCmd.RoutingInfo(0x0, 0x0, 0xc8201a56b6, 0xf, 0x0, 0x0)
        /go/src/github.com/deis/workflow-cli/cmd/routing.go:23 +0x16f
github.com/deis/workflow-cli/cmd.(*DeisCmd).RoutingInfo(0xc82011f940, 0x0, 0x0, 0x0, 0x0)
        <autogenerated>:58 +0xb8
github.com/deis/workflow-cli/parser.routingInfo(0xc82011f930, 0x1, 0x1, 0x7fca49edabe0, 0xc82011f940, 0x0, 0x0)
        /go/src/github.com/deis/workflow-cli/parser/routing.go:58 +0x1e0
github.com/deis/workflow-cli/parser.Routing(0xc82011f930, 0x1, 0x1, 0x7fca49edabe0, 0xc82011f940, 0x0, 0x0)
        /go/src/github.com/deis/workflow-cli/parser/routing.go:34 +0x325
main.Command(0xc82011f930, 0x1, 0x1, 0x24)
        /go/src/github.com/deis/workflow-cli/deis.go:133 +0x1d75
main.main()
        /go/src/github.com/deis/workflow-cli/deis.go:19 +0x60
```

maintenance top-level
```
$ deis maintenance
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4b504f]

goroutine 1 [running]:
panic(0x8ef380, 0xc82000e100)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/deis/workflow-cli/cmd.DeisCmd.MaintenanceInfo(0x0, 0x0, 0xc8201bb476, 0xf, 0x0, 0x0)
        /go/src/github.com/deis/workflow-cli/cmd/maintenance.go:23 +0x16f
github.com/deis/workflow-cli/cmd.(*DeisCmd).MaintenanceInfo(0xc82011f970, 0x0, 0x0, 0x0, 0x0)
        <autogenerated>:43 +0xb8
github.com/deis/workflow-cli/parser.maintenanceInfo(0xc82011f960, 0x1, 0x1, 0x7fa96a929a18, 0xc82011f970, 0x0, 0x0)
        /go/src/github.com/deis/workflow-cli/parser/maintenance.go:59 +0x1e0
github.com/deis/workflow-cli/parser.Maintenance(0xc82011f960, 0x1, 0x1, 0x7fa96a929a18, 0xc82011f970, 0x0, 0x0)
        /go/src/github.com/deis/workflow-cli/parser/maintenance.go:34 +0x325
main.Command(0xc82011f960, 0x1, 0x1, 0x24)
        /go/src/github.com/deis/workflow-cli/deis.go:135 +0x2202
main.main()
        /go/src/github.com/deis/workflow-cli/deis.go:19 +0x60
```

maintenance commands:
```
$ deis maintenance:on
Enabling maintenance mode for k8s-hello-world... Error: EOF
$ deis maintenance:off
Disabling maintenance mode for k8s-hello-world... Error: EOF
```

--- a/cmd/maintenance.go
+++ b/cmd/maintenance.go
@@ -18,10 +18,10 @@ func (d DeisCmd) MaintenanceInfo(appID string) error {
 		return err
 	}
 
-	if *appSettings.Maintenance {
-		d.Println("Maintenance mode is on.")
-	} else {
+	if appSettings.Maintenance == nil || !*appSettings.Maintenance {
 		d.Println("Maintenance mode is off.")
+	} else {
+		d.Println("Maintenance mode is on.")
 	}
 	return nil
 }
@@ -47,7 +47,7 @@ func (d DeisCmd) MaintenanceEnable(appID string) error {
 		return err
 	}
 
-	d.Print("done\n\n")
+	d.Println("done")
 	return nil
 }
 
@@ -72,6 +72,6 @@ func (d DeisCmd) MaintenanceDisable(appID string) error {
 		return err
 	}
 
-	d.Print("done\n\n")
+	d.Println("done")
 	return nil
 }

--- a/cmd/maintenance_test.go
+++ b/cmd/maintenance_test.go
@@ -1,0 +1,120 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/arschles/assert"
+	"github.com/deis/controller-sdk-go/api"
+	"github.com/deis/workflow-cli/pkg/testutil"
+)
+
+func TestMaintenanceInfo(t *testing.T) {
+	t.Parallel()
+	cf, server, err := testutil.NewTestServerAndClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+	var b bytes.Buffer
+	cmdr := DeisCmd{WOut: &b, ConfigFile: cf}
+
+	server.Mux.HandleFunc("/v2/apps/rivendell/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		fmt.Fprintf(w, `{
+			"owner": "elrond",
+			"app": "rivendell",
+			"maintenance": true,
+			"routable": true,
+			"created": "2014-01-01T00:00:00UTC",
+			"updated": "2014-01-01T00:00:00UTC",
+			"uuid": "de1bf5b5-4a72-4f94-a10c-d2a3741cdf75"
+		}`)
+	})
+
+	err = cmdr.MaintenanceInfo("rivendell")
+	assert.NoErr(t, err)
+	assert.Equal(t, b.String(), "Maintenance mode is on.\n", "output")
+
+	server.Mux.HandleFunc("/v2/apps/mordor/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		fmt.Fprintf(w, `{
+			"owner": "sauron",
+			"app": "mordor",
+			"maintenance": false,
+			"routable": true,
+			"created": "2014-01-01T00:00:00UTC",
+			"updated": "2014-01-01T00:00:00UTC",
+			"uuid": "de1bf5b5-4a72-4f94-a10c-d2a3741cdf75"
+		}`)
+	})
+	b.Reset()
+
+	err = cmdr.MaintenanceInfo("mordor")
+	assert.NoErr(t, err)
+	assert.Equal(t, b.String(), "Maintenance mode is off.\n", "output")
+
+	// test that no routable field doesn't trigger a panic
+	server.Mux.HandleFunc("/v2/apps/gondor/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		fmt.Fprintf(w, `{
+			"owner": "aragorn",
+			"app": "gondor",
+			"routable": true,
+			"created": "2014-01-01T00:00:00UTC",
+			"updated": "2014-01-01T00:00:00UTC",
+			"uuid": "de1bf5b5-4a72-4f94-a10c-d2a3741cdf75"
+		}`)
+	})
+	b.Reset()
+
+	err = cmdr.MaintenanceInfo("gondor")
+	assert.NoErr(t, err)
+	assert.Equal(t, b.String(), "Maintenance mode is off.\n", "output")
+}
+
+func TestMaintenanceEnable(t *testing.T) {
+	t.Parallel()
+	cf, server, err := testutil.NewTestServerAndClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+	var b bytes.Buffer
+	cmdr := DeisCmd{WOut: &b, ConfigFile: cf}
+
+	server.Mux.HandleFunc("/v2/apps/lothlorien/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		maintenance := true
+		testutil.AssertBody(t, api.AppSettings{Maintenance: &maintenance}, r)
+		fmt.Fprintf(w, `{}`)
+	})
+
+	err = cmdr.MaintenanceEnable("lothlorien")
+	assert.NoErr(t, err)
+	assert.Equal(t, testutil.StripProgress(b.String()), "Enabling maintenance mode for lothlorien... done\n", "output")
+}
+
+func TestMaintenanceDisable(t *testing.T) {
+	t.Parallel()
+	cf, server, err := testutil.NewTestServerAndClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+	var b bytes.Buffer
+	cmdr := DeisCmd{WOut: &b, ConfigFile: cf}
+
+	server.Mux.HandleFunc("/v2/apps/bree/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		maintenance := false
+		testutil.AssertBody(t, api.AppSettings{Maintenance: &maintenance}, r)
+		fmt.Fprintf(w, `{}`)
+	})
+
+	err = cmdr.MaintenanceDisable("bree")
+	assert.NoErr(t, err)
+	assert.Equal(t, testutil.StripProgress(b.String()), "Disabling maintenance mode for bree... done\n", "output")
+}

--- a/cmd/routing.go
+++ b/cmd/routing.go
@@ -18,7 +18,7 @@ func (d DeisCmd) RoutingInfo(appID string) error {
 		return err
 	}
 
-	if *appSettings.Routable {
+	if appSettings.Routable == nil || *appSettings.Routable {
 		d.Println("Routing is enabled.")
 	} else {
 		d.Println("Routing is disabled.")
@@ -47,7 +47,7 @@ func (d DeisCmd) RoutingEnable(appID string) error {
 		return err
 	}
 
-	d.Print("done\n\n")
+	d.Println("done")
 	return nil
 }
 
@@ -73,6 +73,6 @@ func (d DeisCmd) RoutingDisable(appID string) error {
 		return err
 	}
 
-	d.Print("done\n\n")
+	d.Println("done")
 	return nil
 }

--- a/cmd/routing_test.go
+++ b/cmd/routing_test.go
@@ -1,0 +1,119 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/arschles/assert"
+	"github.com/deis/controller-sdk-go/api"
+	"github.com/deis/workflow-cli/pkg/testutil"
+)
+
+func TestRoutingInfo(t *testing.T) {
+	t.Parallel()
+	cf, server, err := testutil.NewTestServerAndClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+	var b bytes.Buffer
+	cmdr := DeisCmd{WOut: &b, ConfigFile: cf}
+
+	server.Mux.HandleFunc("/v2/apps/rivendell/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		fmt.Fprintf(w, `{
+			"owner": "elrond",
+			"app": "rivendell",
+			"maintenance": true,
+			"routable": true,
+			"created": "2014-01-01T00:00:00UTC",
+			"updated": "2014-01-01T00:00:00UTC",
+			"uuid": "de1bf5b5-4a72-4f94-a10c-d2a3741cdf75"
+		}`)
+	})
+
+	err = cmdr.RoutingInfo("rivendell")
+	assert.NoErr(t, err)
+	assert.Equal(t, b.String(), "Routing is enabled.\n", "output")
+
+	server.Mux.HandleFunc("/v2/apps/mordor/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		fmt.Fprintf(w, `{
+			"owner": "sauron",
+			"app": "mordor",
+			"maintenance": true,
+			"routable": false,
+			"created": "2014-01-01T00:00:00UTC",
+			"updated": "2014-01-01T00:00:00UTC",
+			"uuid": "de1bf5b5-4a72-4f94-a10c-d2a3741cdf75"
+		}`)
+	})
+	b.Reset()
+
+	err = cmdr.RoutingInfo("mordor")
+	assert.NoErr(t, err)
+	assert.Equal(t, b.String(), "Routing is disabled.\n", "output")
+
+	// test that no routable field doesn't trigger a panic
+	server.Mux.HandleFunc("/v2/apps/gondor/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		fmt.Fprintf(w, `{
+			"owner": "aragorn",
+			"app": "gondor",
+			"maintenance": true,
+			"created": "2014-01-01T00:00:00UTC",
+			"updated": "2014-01-01T00:00:00UTC",
+			"uuid": "de1bf5b5-4a72-4f94-a10c-d2a3741cdf75"
+		}`)
+	})
+	b.Reset()
+
+	err = cmdr.RoutingInfo("gondor")
+	assert.NoErr(t, err)
+	assert.Equal(t, b.String(), "Routing is enabled.\n", "output")
+}
+
+func TestRoutingEnable(t *testing.T) {
+	t.Parallel()
+	cf, server, err := testutil.NewTestServerAndClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+	var b bytes.Buffer
+	cmdr := DeisCmd{WOut: &b, ConfigFile: cf}
+
+	server.Mux.HandleFunc("/v2/apps/lothlorien/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		testutil.AssertBody(t, api.AppSettings{Routable: api.NewRoutable()}, r)
+		fmt.Fprintf(w, `{}`)
+	})
+
+	err = cmdr.RoutingEnable("lothlorien")
+	assert.NoErr(t, err)
+	assert.Equal(t, testutil.StripProgress(b.String()), "Enabling routing for lothlorien... done\n", "output")
+}
+
+func TestRoutingDisable(t *testing.T) {
+	t.Parallel()
+	cf, server, err := testutil.NewTestServerAndClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer server.Close()
+	var b bytes.Buffer
+	cmdr := DeisCmd{WOut: &b, ConfigFile: cf}
+
+	server.Mux.HandleFunc("/v2/apps/bree/settings/", func(w http.ResponseWriter, r *http.Request) {
+		testutil.SetHeaders(w)
+		routable := false
+		testutil.AssertBody(t, api.AppSettings{Routable: &routable}, r)
+		fmt.Fprintf(w, `{}`)
+	})
+
+	err = cmdr.RoutingDisable("bree")
+	assert.NoErr(t, err)
+	assert.Equal(t, testutil.StripProgress(b.String()), "Disabling routing for bree... done\n", "output")
+}

